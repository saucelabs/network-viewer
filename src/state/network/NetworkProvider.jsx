import React, { useEffect, useReducer, useMemo } from 'react';
import PropTypes from 'prop-types';

import { reducer, initialState as defaultState } from './reducer';
import { updateData, fetchFile, updateScrollToIndex } from './actions';
import { NetworkContext } from './Context';
import { findRequestIndex } from '../../utils';
import { ROW_ID_PREFIX } from '../../constants';

const NetworkProvider = (props) => {
  const {
    data,
    file,
    fetchOptions,
    initialState,
    scrollTimeStamp,
    scrollRequestPosition,
    autoHighlightChange,
    onDataLoaded,
    onDataError,
    onPause,
    onResume,
    onReset,
  } = props;

  const [state, dispatch] = useReducer(reducer, initialState);
  const callbacks = {
    onPause,
    onResume,
    onReset,
  };
  const value = useMemo(() => [state, dispatch, callbacks], [state]);
  const selectedReqIndex = value[0].get('selectedReqIndex');
  const requestData = value[0].get('data');
  const showReqDetail = value[0].get('showReqDetail');
  const actualData = value[0].get('actualData');
  const error = value[0].get('error');

  // Update data onChange of network data
  useEffect(() => {
    if (data && data.log && data.log.entries) {
      updateData(dispatch)(data);
    }
  }, [data]);

  // Fetch HAR file onChange of file prop
  useEffect(() => {
    if (file) {
      fetchFile(dispatch)(file, fetchOptions);
    }
  }, [file]);

  useEffect(() => {
    if (actualData.size && onDataLoaded) {
      onDataLoaded(actualData);
    }
  }, [actualData]);

  useEffect(() => {
    if (error && onDataError) {
      onDataError(error);
    }
  }, [error]);

  // Find nearby request-rowId and update scrollIndex on scrollTimeStamp receive
  useEffect(() => {
    const shouldChangeHighlight = showReqDetail ? autoHighlightChange : true;
    if (scrollTimeStamp && shouldChangeHighlight) {
      const reqIndex = findRequestIndex({
        data: requestData,
        timestamp: scrollTimeStamp,
        position: scrollRequestPosition,
      });
      if (reqIndex || reqIndex === 0) {
        updateScrollToIndex(dispatch)(requestData.get(reqIndex));
      }
    }
  }, [scrollTimeStamp, actualData]);

  // Scroll to request row onChange of scrollToIndex
  // setTimeout is required when reqDetail is visible,
  // it allows DOM to adjust before we scroll it to highlighted request
  useEffect(() => {
    if (selectedReqIndex) {
      setTimeout(() => {
        const element = document.getElementById(ROW_ID_PREFIX + selectedReqIndex);
        if (element) {
          element.scrollIntoView({
            alignToTop: true,
            behavior: 'smooth',
          });
        }
      }, 300);
    }
  }, [selectedReqIndex]);

  return (
    <NetworkContext.Provider
      value={value}
      {...props}
    />
  );
};

NetworkProvider.propTypes = {
  autoHighlightChange: PropTypes.bool,
  data: PropTypes.object,
  fetchOptions: PropTypes.object,
  file: PropTypes.string,
  initialState: PropTypes.object,
  onDataError: PropTypes.func,
  onDataLoaded: PropTypes.func,
  onPause: PropTypes.func,
  onReset: PropTypes.func,
  onResume: PropTypes.func,
  scrollRequestPosition: PropTypes.oneOf(['before', 'after', 'near']),
  scrollTimeStamp: PropTypes.number,
};

NetworkProvider.defaultProps = {
  autoHighlightChange: false,
  data: null,
  fetchOptions: { withCredentials: true },
  file: null,
  initialState: defaultState,
  onDataError: null,
  onDataLoaded: null,
  onPause: null,
  onReset: null,
  onResume: null,
  scrollRequestPosition: 'near',
  scrollTimeStamp: null,
};

export default NetworkProvider;
