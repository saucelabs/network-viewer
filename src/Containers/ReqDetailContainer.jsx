import React from 'react';

import { useNetwork } from './../state/network/Context';
import Styles from './ReqDetailContainer.styles.scss';
import Tabs from '../Components/Common/Tabs';
import Tab from '../Components/Common/Tab';
import Headers from './../Components/ReqDetail/Headers';
import IconCloseSign from './../icons/IconCloseSign';
import Response from '../Components/ReqDetail/Response';

const ReqDetailContainer = () => {
  const { actions, state } = useNetwork();
  const reqDetail = state.get('reqDetail');
  const handleCloseClick = () => {
    actions.selectRequest(null);
  };

  return (
    <div className={Styles['req-detail-container']}>
      <button
        className={Styles['close-button']}
        onClick={handleCloseClick}
        type="button"
      >
        <IconCloseSign className={Styles['close-icon']} />
      </button>
      <Tabs
        activeClassName={Styles.active}
        defaultSelectedKey="headers"
        navLinkClassName={Styles['tab-link']}
        navTabsClassName={Styles['nav-tabs']}
        tabsContainerClassName={Styles['tabs-container']}
      >
        <Tab
          eventKey="headers"
          name="Headers"
        >
          <Headers data={reqDetail} />
        </Tab>
        <Tab
          eventKey="response"
          name="Response"
        >
          <Response data={reqDetail} />
        </Tab>
      </Tabs>
    </div>
  );
};

export default ReqDetailContainer;
