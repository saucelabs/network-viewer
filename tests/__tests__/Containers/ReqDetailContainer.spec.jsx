import React from 'react';
import { mount } from 'enzyme';

import ReqDetailContainer from './../../../src/Containers/ReqDetailContainer';
import NetworkProvider from './../../../src/state/network/NetworkProvider';

describe('ReqDetailContainer', () => {
  it('renders without crashing', () => {
    const element = mount(
      <NetworkProvider>
        <ReqDetailContainer />
      </NetworkProvider>,
    );
    expect(element).toMatchSnapshot();
  });
});
