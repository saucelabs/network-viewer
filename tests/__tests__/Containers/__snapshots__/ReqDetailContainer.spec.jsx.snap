// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReqDetailContainer renders without crashing 1`] = `
<NetworkProvider
  autoHighlightChange={false}
  data={null}
  fetchOptions={
    Object {
      "withCredentials": true,
    }
  }
  file={null}
  initialState={
    Immutable.Map {
      "errorFilter": false,
      "search": Object {
        "name": "url",
        "value": "",
      },
      "error": null,
      "scrollToIndex": null,
      "data": Immutable.List [],
      "selectedReqIndex": null,
      "reqDetail": null,
      "showReqDetail": false,
      "totalNetworkTime": null,
      "actualData": Immutable.List [],
      "filter": Object {
        "name": null,
        "value": null,
      },
      "dataSummary": Immutable.Map {},
      "loading": false,
      "sort": Object {
        "isAcs": true,
        "key": "startedDateTime",
      },
    }
  }
  onDataError={null}
  onDataLoaded={null}
  scrollRequestPosition="near"
  scrollTimeStamp={null}
>
  <ReqDetailContainer>
    <div>
      <button
        aria-label="Close button"
        onClick={[Function]}
        type="button"
      >
        <IconCloseSign
          className=""
        >
          <svg
            className=""
            height="16"
            viewBox="0 0 17 16"
            width="17"
          >
            <g
              fill="none"
              fillRule="evenodd"
              stroke="#ACB2B8"
              strokeLinecap="round"
              transform="translate(1)"
            >
              <path
                d="M0 0L15.2460858 15.2460858M15.2460858 0L0 15.2460858"
              />
            </g>
          </svg>
        </IconCloseSign>
      </button>
      <Tabs
        activeClassName={null}
        defaultSelectedKey="headers"
        navLinkClassName={null}
        navTabsClassName={null}
        onUpdate={null}
        selectedKey={null}
        tabsContainerClassName={null}
      >
        <nav
          className="nav-tabs"
        >
          <a
            className="tab-item undefined"
            key="headers"
            onClick={[Function]}
            role="tab"
            tabIndex={0}
          >
            Headers
          </a>
          <a
            className="tab-item"
            key="response"
            onClick={[Function]}
            role="tab"
            tabIndex={1}
          >
            Response
          </a>
        </nav>
        <section
          className=""
        >
          <Headers
            data={null}
          />
        </section>
      </Tabs>
    </div>
  </ReqDetailContainer>
</NetworkProvider>
`;
